backend/app/services/ObjectStorage.scala

```diff
@@ -38,8 +39,8 @@ class S3ObjectStorage private(client: S3Client, bucket: String) extends ObjectSt
     run(client.aws.deleteObject(bucket, key))
   }
 
-  def delete(keys: List[String]): Either[Failure, Unit] = {
-    val request = new DeleteObjectsRequest(bucket).withKeys(keys: _*)
+  def deleteMultiple(keys: Set[String]): Either[Failure, Unit] = {
+    val request = new DeleteObjectsRequest(bucket).withKeys(keys.toSeq: _*)
     run(client.aws.deleteObjects(request))
```

Comment:
This is probably not crucial, but I thought it was worth noting that a batch command like this can fail to delete some keys, but still return a response that is 'successful' in terms of HTTP response code. The 'things that failed' are [buried inside the response](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html#API_DeleteObjects_Examples:~:text=Object%3E%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%3C/Delete%3E-,Sample%20Response,-The%20response%20includes), eg:

```xml
HTTP/1.1 200 OK
            x-amz-id-2: 5h4FxSNCUS7wP5z92eGCWDshNpMnRuXvETa4HH3LvvH6VAIr0jU7tH9kM7X+njXx
            x-amz-request-id: A437B3B641629AEE
            Date: Fri, 02 Dec 2011 01:53:42 GMT
            Content-Type: application/xml
            Server: AmazonS3
            Content-Length: 251

            <?xml version="1.0" encoding="UTF-8"?>
            <DeleteResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
             <Deleted>
               <Key>sample1.txt</Key>
             </Deleted>
             <Error>
              <Key>sample2.txt</Key>
              <Code>AccessDenied</Code>
              <Message>Access Denied</Message>
             </Error>
            </DeleteResult>
```

...in this context, it might mean that the deleteMultiple() method should return failure if even one object fails to be deleted, and I'm not sure if the code would currently do this. However, as partial failure (eg due to access denied for just some objects, and not others) is unlikely to happen, this might not be a case that needs urgently handling.

Reply:
Great point. I think for the sake of our pseudo-transaction (do not delete the elasticsearch document until we're sure all other traces in S3 and neo4j are gone), it would probably be best to fail the whole thing if any individual deletion fails. However, we also don't want to consider "object does not exist" a failure because sometimes we try to delete previews where there is none, and that's fine. So that might complicate it...

Reply:
Actually, the [Java SDK docs](https://docs.aws.amazon.com/zh_tw/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3.html#deleteObjects-com.amazonaws.services.s3.model.DeleteObjectsRequest-) say "If any object in the request cannot be deleted, this method throws a [MultiObjectDeleteException](https://docs.aws.amazon.com/zh_tw/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/MultiObjectDeleteException.html) with details of the error", so we should get a `Left` and fail as we would like.

Reply:
I've also just quickly verified by testing that [this comment](https://github.com/guardian/giant/blob/main/backend/app/commands/DeleteResource.scala#L59) is still correct for multiple deletions. A multiple delete that includes one object that does not exist succeeds.

Reply:
Ah fantastic, nice one!